src/v/datalake/schema_registry.cc

```diff
@@ -24,4 +25,36 @@ get_schema_id_result get_value_schema_id(const iobuf& buf) {
     return pandaproxy::schema_registry::schema_id{id};
 }
 
+// TODO: this is mostly a copy-and-paste of get_proto_offsets from
+// pandaproxy::schema_registry with a slightly different interface. Unify these.
+get_proto_offsets_result get_proto_offsets(const iobuf& buf) {
+    // The encoding is a length, followed by indexes into the file or message.
+    // Each number is a zigzag encoded integer.
+    iobuf_const_parser parser(buf);
+    // skip schema id prefix
+    parser.skip(sizeof(uint8_t) + sizeof(int32_t));
+    std::vector<int32_t> offsets;
+    auto [offset_count, bytes_read] = parser.read_varlong();
+    if (!bytes_read) {
+        return get_schema_error::bad_varint;
+    }
+    // Reject more offsets than bytes remaining; it's not possible
+    if (static_cast<size_t>(offset_count) > parser.bytes_left()) {
+        return get_schema_error::not_enough_bytes;
+    }
+    offsets.resize(offset_count);
+    for (auto& o : offsets) {
+        std::tie(o, bytes_read) = parser.read_varlong();
```

Comment:
doesn't this provide an inconsistent semantic for short read scenarios? for example, even if offset_count < parser.bytes_bytes() then you could still run out of bytes when parsing since some logical offsets may take more than 1 byte read from the parser? in this case, you could get back not_enough_bytes or some exception thrown out of the parser for the same error scenario?

Reply:
This is a bit inconsistent.

There are a few different reasons we might stop parsing a varint:
- Found a byte with the top bit cleared.
- Read the maximum number of bytes for the given type.
- Got to the end of the buffer.

Our varint parser doesn't distinguish between those cases. E.g. `detail::var_decoder` in vint.h returns true (meaning "stop parsing") upon finding a byte with the top bit cleared, or when "shift > limit", i.e. the number of bits read is greater than the size.

`deserialize` in that same module loops through a range until it runs out of bytes or val_decoder returns true and returns the number of bytes read.

As far as I can tell, currently the only way we could return 0 bytes read is if we call it at the end of the buffer. Otherwise it always reads at least on byte. Also, we currently don't detect improperly formatted varints. So maybe I should just return `not_enough_bytes` in both cases.

In the future we could detect improperly-formatted varints and return 0 bytes read from the parser in those cases. In that case it would definitely make sense to separate these cases.

Reply:
thanks! i took a closer look at our varint decoder. looks like we have some improvements to it we can make in the future!